### PR TITLE
Feat/geocoding

### PR DESCRIPTION
**Pull request by e-bernier**
This **PR** introduces the initial geocoding functionality using the OpenStreetMap Nominatim API.
It includes core models, repository logic, rate limiting, and unit tests.

What's Included:
- Location model: Represents a geographic point with name, latitude, and longitude.
- LocationRepository interface: Defines a coroutine-based contract for location search.
- NominatimLocationRepository: Implements location search with caching, rate limiting, and JSON parsing.
- RateLimiter utility: Ensures API requests respect minimum intervals.
- LocationSearchException: Custom exception for search failures.
- Unit tests: Covers parsing, caching, rate limiting, and error handling using mockK.
- Build config updates: Adds mockK and org.json dependencies for testing.

### CODE REVIEW

File: app/src/main/java/com/swent/mapin/model/NominatimLocationRepository.kt [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+/**
+ * A repository implementation that uses the Nominatim API to search for geographic locations.
+ *
+ * This class performs network requests to the OpenStreetMap Nominatim service to retrieve location
+ * data based on a query string. It includes basic caching and rate-limiting to reduce redundant
+ * requests and avoid hitting API limits.
+ *
+ * @property client An instance of [OkHttpClient] used to perform HTTP requests.
+ */
```
Comment (lyesehili) [inline app/src/main/java/com/swent/mapin/model/NominatimLocationRepository.kt:24]: Documentation is very clear!

File: app/src/test/java/com/swent/mapin/model/NominatimLocationRepositoryTests.kt [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+ * - Enforcing rate limits between API calls
+ * - Handling successful and failed HTTP responses
+ * - Using cached results for repeated queries
+ * - Skipping invalid location entries
+ * - Throwing exceptions for null or failed responses
+ *
+ * Mocks are used to simulate OkHttp client behavior and isolate repository logic.
+ */
+@OptIn(ExperimentalCoroutinesApi::class)
+class NominatimLocationRepositoryTests {
```
Comment (lyesehili) [inline app/src/test/java/com/swent/mapin/model/NominatimLocationRepositoryTests.kt:39]: Great coverage of edge cases in your test suite.

**Review by lyesehili — APPROVED** [review]
This PR adds a well-structured location search feature using the Nominatim API. The code is clean, well-documented, and tested. All new dependencies are logically added and organized. Consider future-proofing the rate limiter for concurrent access if needed, but overall, this is a strong and maintainable addition to the codebase. Great work!
